x-pack/plugins/dashboard_enhanced/public/services/drilldowns/dashboard_to_dashboard_drilldown/drilldown.test.tsx

```diff
@@ -4,17 +4,335 @@
+          query: { query: 'test', language: 'kuery' },
+          ...embeddableInput,
+        }),
+      },
+    } as unknown) as ActionContext<VisualizeEmbeddableContract>;
+
+    await drilldown.execute(completeConfig, context);
+
+    if (useRangeEvent) {
+      expect(selectRangeFiltersSpy).toBeCalledTimes(1);
+      expect(valueClickFiltersSpy).toBeCalledTimes(0);
+    } else {
+      expect(selectRangeFiltersSpy).toBeCalledTimes(0);
+      expect(valueClickFiltersSpy).toBeCalledTimes(1);
+    }
+
+    expect(navigateToApp).toBeCalledTimes(1);
+    expect(navigateToApp.mock.calls[0][0]).toBe('kibana');
+
+    const executeNavigatedPath = navigateToApp.mock.calls[0][1]?.path;
+    const href = await drilldown.getHref(completeConfig, context);
+
+    expect(href.includes(executeNavigatedPath)).toBe(true);
+
+    return {
+      href,
+    };
+  }
+
+  test('navigates to correct dashboard', async () => {
+    const testDashboardId = 'dashboardId';
+    const { href } = await setupTestBed(
+      {
+        dashboardId: testDashboardId,
+      },
+      {},
+      [],
+      false
+    );
+
+    expect(href).toEqual(expect.stringContaining(`dashboard/${testDashboardId}`));
+  });
+
+  test('query is removed with query if filters are disabled', async () => {
+    const queryString = 'querystring';
+    const queryLanguage = 'kuery';
+    const { href } = await setupTestBed(
+      {
+        useCurrentFilters: false,
+      },
+      {
+        query: { query: queryString, language: queryLanguage },
+      },
+      []
+    );
+
+    expect(href).toEqual(expect.not.stringContaining(queryString));
+    expect(href).toEqual(expect.not.stringContaining(queryLanguage));
+  });
+
+  test('navigates with query if filters are enabled', async () => {
+    const queryString = 'querystring';
+    const queryLanguage = 'kuery';
+    const { href } = await setupTestBed(
+      {
+        useCurrentFilters: true,
+      },
+      {
+        query: { query: queryString, language: queryLanguage },
+      },
+      []
+    );
+
+    expect(href).toEqual(expect.stringContaining(queryString));
+    expect(href).toEqual(expect.stringContaining(queryLanguage));
+  });
+
+  test('when user chooses to keep current filters, current filters are set on destination dashboard', async () => {
+    const existingAppFilterKey = 'appExistingFilter';
+    const existingGlobalFilterKey = 'existingGlobalFilter';
+    const newAppliedFilterKey = 'newAppliedFilter';
+
+    const { href } = await setupTestBed(
+      {
+        useCurrentFilters: true,
+      },
+      {
+        filters: [getFilter(false, existingAppFilterKey), getFilter(true, existingGlobalFilterKey)],
+      },
+      [getFilter(false, newAppliedFilterKey)]
+    );
+
+    expect(href).toEqual(expect.stringContaining(existingAppFilterKey));
+    expect(href).toEqual(expect.stringContaining(existingGlobalFilterKey));
+    expect(href).toEqual(expect.stringContaining(newAppliedFilterKey));
```

Comment:
Are filters that are saved with the destination dashboard preserved?  There are three filters: dynamic filters, filters on source dashboard, and filters on destination dashboard.  If you say yes to filters on current dashboard, will the final filters be all three or are the destination filters replaced?

Reply:
Just checked. Looks like no. If there is `_a` in the dashboard url, then that state takes precedence. 
And this is tricky. I imagine If I share a url to dashboard, I'd expect receiver to see the same filters, which I am seeing 🤔. And this how it is right now. - so IMO dashboard restoring state from the URL is correct.

I was trying to avoid it, but looks like we will have to pull in `destination` dashboard inside `dashboard_to_dashboard_drilldown` to its get saved filters and to put them into destination URL, correct? 
- Good thing, this will also improve error handling for cases when `destination` dashboard no longer exist and will not navigate user to not existing dashboard. 

Agree?

Reply:
Yep, that's what I was thinking as well.

Reply:
@stacey-gammon, thinking a bit more about it: 

Initially I though to pull in destination dashboard inside the `dashboard_drilldown` code and pass merged filters into url generator. 
But know I am realising the this is likely most common scenario? Whoever uses dashboard url generator wants to preserve original filters also. And it seems it would be too much burden for every consumer to retrieve saved filters before using url generator. 
So options:

1. Option 1: retrieve saved filters in `Drilldowns` code. Pass all filters to url generator
2. Option 2: retrieve saved filters inside url generator. merge them with incoming filters when generating url
3. Option 3: Url generator optionally adds additional flag in the url to also preserve saved filters: e.g. `&savedFilters=merge`. When dashboard is loaded it will use that flag to decided what to do with filters saved in dashboard. 

I am leaning towards `Option 3`, because less dependencies inside url generator (no need to pull in saved object) and generated url does not contain snapshot of filters, but will retrieve the ones which are saved with dashboard when dashboard is opened.

What do you think?

Reply:
Option 3 sounds okay, we should ask @ThomThomson and @majagrubic to weigh in though since ultimately this is their code to maintain. 

--  brain dump --

I am really a bit torn on Option 2 vs 3.  Where do we want the bulk of the logic to reside - url generators or apps handling a slew of different url parameters?  If you keep your URL parameters to a minimum, it feels easier to maintain and migrate. We don't have a version number stored with the URL, but we do with the url generator state (via the id, though we probably should change it to accept an actual version number).  

Say we wanted to split `savedFilters=merge` into two parameters, one for the query and one for filter pills.  There is a clear path forward for how to do that with URL generators. The migration logic is encapsulated in one place.  Dashboard app code doesn't get littered with a bunch of extra "ifs" to try to figure out what URL parameters are in there.  There is no chance of having to figure out how to handle a URL mixed with legacy and the new parameters.

I don't think we have a good story, outside of URL generators, for deprecating/migrating urls. I might be wrong though, if this is where state containers comes into play. Every time I have a conversation with @lukeelmers about url generators vs state containers, I feel like I am lacking a solid understanding of the state containers.

So, I think we can move forward either way without blocking ourselves in, but it's on my todo list to try to clarify this problem of URL migrations.

Reply:
> Url generator optionally adds additional flag in the url to also preserve saved filters: e.g. `&savedFilters=merge`. When dashboard is loaded it will use that flag to decided what to do with filters saved in dashboard.

UX question - Is there ever a scenario in dashboard `view` mode where we _wouldn't_ want to preserve saved filters & merge in any incoming ones which are new? I agree with @Dosant that it is reasonable for a user saving a dashboard with filters to expect those filters to be preserved when anybody returns to the dashboard. So I guess I'm wondering why we'd need to add a `savedFilters=merge` param at all, vs making this the standard behavior for URLs in the dashboard app.

If a dashboard owner saves it with filters, I think it is an acceptable tradeoff that those filters cannot be wiped out by new filters coming in from a drilldown. I would expect you to be able to add new ones, but not fundamentally alter the existing dashboard.

Edit mode might be a different story, IDK. Perhaps there's a use case for overriding saved filters in that case since they are being edited.

--- everything below is a tangent based on stacey's brain dump above ---

> I don't think we have a good story, outside of URL generators, for deprecating/migrating urls.

My two cents on this is that things are going to become overly complex if the concept of url state/migrations somehow diverges from the concept of persistable state/migrations in general. IMO the proposed [persistable state registry](https://github.com/elastic/kibana/issues/62950) & the idea for [persistable state migrations](https://github.com/elastic/kibana/issues/63358) should also serve as the solution to dealing with migrations in urls. 

So in that regard, I don't think we need a migration solution which is specific to URLs in the long run: The URLs are a reflection of an application's internal persistable state, not the single source of truth for state, so when we talk about migration of URLs, what we are actually talking about is migration of state.

Most likely, this means either removing migrations as a concept from the URL generators, or having authors of those migrations use persistable states registry under the hood. But I 💯 agree with @stacey-gammon that we need to work to articulate a clearer story of how state migrations apply to URLs more broadly. This should be a design consideration for the persistable states registry work.

Reply:
>  Is there ever a scenario in dashboard view mode where we wouldn't want to preserve saved filters & merge in any incoming ones which are new?

I think the situation we need to be sure to handle is:

- User opens dashboard 123 with saved filter: `country:US`
- User removes `country:US` filter, in _view_ mode.

That is handled by using the `filters` in the URL as the source of truth.  I don't see how we could represent that state, in view mode, if it was only allowed to be additive.

This makes me lean more towards Option 2 as a better route because of these subtleties.   If you open up a dashboard with `savedFilters=merge`, will dashboard load up, merge the filters, and _remove_ that parameter?  If it stays in the URL, how could the user then delete a filter in view mode?

> IMO the proposed persistable state registry & the idea for persistable state migrations should also serve as the solution to dealing with migrations in urls.

100% agree with this.  Adding this URL specific migration code was me experimenting with migrations in a pre-persistable state registry world, knowing that as soon as a generic solution was available, URL generators would use that instead of it's own system.

Reply:
@stacey-gammon,

> This makes me lean more towards Option 2 as a better route because of these subtleties. If you open up a dashboard with savedFilters=merge, will dashboard load up, merge the filters, and remove that parameter?

Yes, I was planning to remove the parameter using `history.replace` when dashboard is loaded.

What I don't like about `Option2` is that we will have to handle inside url generator scenario when dashboard with id is missing. Or handle any other error where we wasn't able to load the saved object. Should we in this case just swallow the error and don't add any additional filters into the url? Or should we throw an error, so don't generate any url? 
I'd prefer to try to not to do any of that inside the url generator and leave it to dashboard app instead

Reply:
> What I don't like about Option2 is that we will have to handle inside url generator scenario when dashboard with id is missing.  Or handle any other error where we wasn't able to load the saved object. Should we in this case just swallow the error and don't add any additional filters into the url? 

🤔 

hmm, what is the proper way to handle errors in these url generators is a good question. You could return the dashboard listing page url, though you'd have to manually pop up the "missing dashboard" notification. So I'd probably just swallow the error and not add any filters, and return the url to the dashboard with the id which would then just redirect to the landing page and show the "missing dashboard" notification.  But I admit, I don't _love_ that flow.  

I still think that there is more overhead and exposure in adding this to the URL, which is a public contract that needs to be maintained and tested.  We were just talking in the tech lead sync today about URLs and how, if we want to strive to get to a place where all the address bar URLs are BWC for minimum one major, ideally longer, we should be putting less information in them.

It's easy to write unit tests for this URL generator based on different input parameters (what happens if the id is missing), but there is no way easy way to write unit tests for all the url parameters dashboard app should handle. You'd have to write a functional test to cover it, and those are slower, and you'd need to write a lot of them to cover everything (what happens if I use a non-existant id _and_ have this new parameter set to true, set to false, what if I use this setting in view mode, in edit mode, etc).

Ultimately I leave it up to the final maintainers of this code, which are @ThomThomson and @majagrubic.

Reply:
@stacey-gammon, fyi

@majagrubic, also considers Option 2 better: https://github.com/elastic/kibana/pull/64633#issuecomment-621115789

So I went with Option 2 here: https://github.com/elastic/kibana/pull/64767